src/worker/jobs/expiry_notification.rs

```diff
@@ -0,0 +1,188 @@
+use crate::{email::Email, models::User, worker::Environment, Emails};
+use anyhow::anyhow;
+use crates_io_worker::BackgroundJob;
+use diesel::{
+    dsl::now, Connection, ExpressionMethods, NullableExpressionMethods, PgConnection, RunQueryDsl,
+};
+use std::sync::Arc;
+
+/// The threshold in days for the expiry notification.
+const EXPIRY_THRESHOLD: i64 = 3;
+
+/// A job responsible for monitoring the status of a token.
+/// It checks if the token is about to reach its expiry date.
+/// If the token is about to expire, the job triggers a notification.
+#[derive(Default, Serialize, Deserialize, Debug)]
+pub struct CheckAboutToExpireToken;
+
+impl BackgroundJob for CheckAboutToExpireToken {
+    const JOB_NAME: &'static str = "expiry_notification";
+
+    type Context = Arc<Environment>;
+
+    #[instrument(skip(env), err)]
+    async fn run(&self, env: Self::Context) -> anyhow::Result<()> {
+        let mut conn = env.connection_pool.get()?;
+        // Check if the token is about to expire
+        // If the token is about to expire, trigger a notification.
+        check(&env.emails, &mut conn)
+    }
+}
+// Check if the token is about to expire and send a notification if it is.
+fn check(emails: &Emails, conn: &mut PgConnection) -> anyhow::Result<()> {
+    info!("Checking if tokens are about to expire");
+    let expired_tokens =
+        crate::models::token::ApiToken::find_tokens_expiring_within_days(conn, EXPIRY_THRESHOLD)?;
+    // Batch send notifications in transactions.
+    const BATCH_SIZE: usize = 100;
+    for chunk in expired_tokens.chunks(BATCH_SIZE) {
```

Comment:
why are we chunking this? and if we decide to chunk it, why not chunk it at the database level?

Reply:
I am more concerned here about having too many tokens that need to be emailed, resulting in a large transaction. I may be overly worried?

Reply:
I think what @Turbo87 means is why not just select data in batches (`select ... limit BATCH_SIZE`) instead of selecting all the data at once and then chunking it?

Reply:
> if e.g. we process 10 tokens and the 7th update fails, we have already sent out 6 emails, but since the transaction will be rolled back we wouldn't have saved that in the database

In the current implementation, we just print a log if the send is failed. We will try it in the next round.

Reply:
> why not just select data in batches (`select ... limit BATCH_SIZE`) instead of selecting all the data at once and then chunking it?

Is it possible that after an email fails to send, we will immediately retry it again in the next loop? (This relies on the database implementation of limit) Would this cause us to try it many times in a short period of time.
Or do you mean we order the data and use limit + offset to control this?

Reply:
Yeah, if we need to run in batch mode, then avoiding errors that cause rollbacks would be more reasonable.

On the other hand, if we process emails one at a time, then we need a mechanism to prevent getting stuck selecting the failed email, which would block other unhandled emails.

Reply:
> Is it possible that after an email fails to send, we will immediately retry it again in the next loop? (This relies on the database implementation of limit) Would this cause us to try it many times in a short period of time. Or do you mean we order the data and use limit + offset to control this?

I think this part is similar to how we do seek(cursor)-based pagination. We just need to use `ORDER BY` to ensure a deterministic order and seek columns for offset.
e.g., `SELECT ... WHERE id > last_processed_id ORDER BY id LIMIT batch_size`.
P.S. There's no real need to order by ID, we could also order by expiry time if that makes more sense.

Reply:
> Yeah, if we need to run in batch mode, then avoiding errors that cause rollbacks would be more reasonable.

Expect the email not found case. We already did it.

Reply:
> On the other hand, if we process emails one at a time, then we need a mechanism to prevent getting stuck selecting the failed email, which would block other unhandled emails.

Yes, I think sending them one by one makes it more complex. It's difficult to avoid we stuck on the same one for a while.

Reply:
@eth3lbert @Turbo87 Based on the comments mentioned above, how about we continue using the current approach and log any errors for emails that failed or were not found? Afterward, we can proceed to send all other emails that can be sent. When we trigger the background job again, we can attempt to send the failed emails again. Does this sound reasonable to you?

Using batch + offset or sending them one by one seems all require more effort to avoid retrying the same email in a short period of time. I am also OK with the `batch + offset` solution. I'm not sure how many expired tokens are in crates.io right now. Maybe I'm overthinking and complicating the issue. 😆

Reply:
> Yes, I think sending them one by one makes it more complex. It's difficult to avoid we stuck on the same one for a while.

The following is a simplified version for sending them one by one.

```rust
fn get_token(
    last_id: Option<i32>,
    conn: &mut PgConnection,
) -> diesel::QueryResult<crate::models::token::ApiToken> {
    use crate::schema::api_tokens;
    use diesel::prelude::*;

    let now = chrono::Utc::now();
    let expired_at = now + chrono::TimeDelta::try_days(EXPIRY_THRESHOLD).unwrap();

    let mut query = api_tokens::table
        .filter(api_tokens::revoked.eq(false))
        .filter(
            api_tokens::expired_at
                .is_not_null()
                .and(api_tokens::expired_at.assume_not_null().gt(now.naive_utc())) // Ignore already expired tokens
                .and(
                    api_tokens::expired_at
                        .assume_not_null()
                        .lt(expired_at.naive_utc()),
                ),
        )
        .into_boxed();

    if let Some(id) = last_id {
        query = query.filter(api_tokens::id.gt(id));
    }

    query
        .order_by(crate::schema::api_tokens::id.asc())
        .select(crate::models::token::ApiToken::as_select())
        .filter(api_tokens::expiry_notification_at.is_null())
        .first::<crate::models::token::ApiToken>(conn)
}

// Check if the token is about to expire and send a notification if it is.
fn check(emails: &Emails, conn: &mut PgConnection) -> anyhow::Result<()> {
    info!("Checking if tokens are about to expire");

    let mut last_id: Option<i32> = None;

    // You might want to consider changing the following loop to a while let loop
    loop {
        let Ok(token) = get_token(last_id, conn) else {
            break;
        };
        last_id = Some(token.id);

        conn.transaction(|conn| {
            let user = User::find(conn, token.user_id)?;
            let Some(recipient) = user.email(conn)? else {
                error!("No address found for token {}", token.id);
                return Ok::<_, anyhow::Error>(());
            };
            let email = ExpiryNotificationEmail {
                name: &user.gh_login,
                token_name: &token.name,
            };
            match emails.send(&recipient, email) {
                Ok(_) => {
                    // Update the token to prevent duplicate notifications.
                    diesel::update(&token)
                        .set(
                            crate::schema::api_tokens::expiry_notification_at
                                .eq(chrono::Utc::now().naive_utc()),
                        )
                        .execute(conn)?;
                }
                Err(e) => {
                    error!("Failed to send email: {:?} to {}", e, recipient);
                }
            }
            Ok(())
        })?;
    }

    Ok(())
}
```

Feel free to modify it for better readability or clarity.

Reply:
I thought a bit more about this and I think the simplest implementation would be to get one big chunk from the database (`LIMIT 10000` or something like that) and then process them one by one without chunking. That would allow us to get rid of the outer `for chunk in chunks` loop and the `conn.transaction()` call since there is only a single database call happening inside each loop iteration. we could potentially add a warning message to the logs if we actually receive the 10000 tokens in any run, but with the current volume such a limit will probably be sufficient.

Reply:
> I thought a bit more about this and I think the simplest implementation would be to get one big chunk from the database (`LIMIT 10000` or something like that) and then process them one by one without chunking.

I have followed this approach and updated the PR.